ICE: incr comp: `Span must not be empty and have no suggestion`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
struct Foo<T>;

macro_rules! foo_ty {
    ($a:ty, $b:ty) => {
        Foo<a, $b>
    }
}
fn foo<'a, 'b>() -> foo_ty!(&'b (), &'b ()) {}

fn main() {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.93.0-nightly (63b1db058 2025-12-02)
binary: rustc
commit-hash: 63b1db05801271e400954e41b8600a3cf1482363
commit-date: 2025-12-02
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5
```

required debug assertions, only reproducible with incr comp (don't ask why..)
`rustc code.rs -Cincremental=.`
### Error output

```
error[E0412]: cannot find type `a` in this scope
 --> code.rs:5:13
  |
5 |         Foo<a, $b>
  |             ^ not found in this scope
...
8 | fn foo<'a, 'b>() -> foo_ty!(&'b (), &'b ()) {}
  |                     ----------------------- in this macro invocation
  |
  = note: this error originates in the macro `foo_ty` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you might be missing a type parameter
  |
8 | fn foo<'a, 'b, a>() -> foo_ty!(&'b (), &'b ()) {}
  |              +++

```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' (3343042) panicked at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_errors/src/diagnostic.rs:1030:9:
Span must not be empty and have no suggestion
stack backtrace:
   0:     0x7f3f3594bd80 - std[bebcfa26ff680867]::backtrace_rs::backtrace::libunwind::trace
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/../../backtrace/src/backtrace/libunwind.rs:117:9
   1:     0x7f3f3594bd80 - std[bebcfa26ff680867]::backtrace_rs::backtrace::trace_unsynchronized::<std[bebcfa26ff680867]::sys::backtrace::_print_fmt::{closure#1}>
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/../../backtrace/src/backtrace/mod.rs:66:14
   2:     0x7f3f3594bd80 - std[bebcfa26ff680867]::sys::backtrace::_print_fmt
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/sys/backtrace.rs:68:9
   3:     0x7f3f3594bd80 - <<std[bebcfa26ff680867]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[91c473b3610db379]::fmt::Display>::fmt
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/sys/backtrace.rs:38:26
   4:     0x7f3f3131fa2e - <core[91c473b3610db379]::fmt::rt::Argument>::fmt
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/fmt/rt.rs:152:76
   5:     0x7f3f3131fa2e - core[91c473b3610db379]::fmt::write
   6:     0x7f3f35964b4c - std[bebcfa26ff680867]::io::default_write_fmt::<std[bebcfa26ff680867]::sys::stdio::unix::Stderr>
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/io/mod.rs:639:11
   7:     0x7f3f35964b4c - <std[bebcfa26ff680867]::sys::stdio::unix::Stderr as std[bebcfa26ff680867]::io::Write>::write_fmt
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/io/mod.rs:1994:13
   8:     0x7f3f3590dc86 - <std[bebcfa26ff680867]::sys::backtrace::BacktraceLock>::print
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/sys/backtrace.rs:41:9
   9:     0x7f3f3590dc86 - std[bebcfa26ff680867]::panicking::default_hook::{closure#0}
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:292:27
  10:     0x7f3f35937221 - std[bebcfa26ff680867]::panicking::default_hook
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:319:9
  11:     0x7f3f32562814 - <alloc[a8d04eb2314326e5]::boxed::Box<dyn for<'a, 'b> core[91c473b3610db379]::ops::function::Fn<(&'a std[bebcfa26ff680867]::panic::PanicHookInfo<'b>,), Output = ()> + core[91c473b3610db379]::marker::Send + core[91c473b3610db379]::marker::Sync> as core[91c473b3610db379]::ops::function::Fn<(&std[bebcfa26ff680867]::panic::PanicHookInfo,)>>::call
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/alloc/src/boxed.rs:2220:9
  12:     0x7f3f32562814 - rustc_driver_impl[13cf1191ca3affe6]::install_ice_hook::{closure#1}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_driver_impl/src/lib.rs:1432:17
  13:     0x7f3f32562814 - <alloc[a8d04eb2314326e5]::boxed::Box<rustc_driver_impl[13cf1191ca3affe6]::install_ice_hook::{closure#1}> as core[91c473b3610db379]::ops::function::Fn<(&dyn for<'a, 'b> core[91c473b3610db379]::ops::function::Fn<(&'a std[bebcfa26ff680867]::panic::PanicHookInfo<'b>,), Output = ()> + core[91c473b3610db379]::marker::Send + core[91c473b3610db379]::marker::Sync, &std[bebcfa26ff680867]::panic::PanicHookInfo)>>::call
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/alloc/src/boxed.rs:2220:9
  14:     0x7f3f32562814 - std[bebcfa26ff680867]::panicking::update_hook::<alloc[a8d04eb2314326e5]::boxed::Box<rustc_driver_impl[13cf1191ca3affe6]::install_ice_hook::{closure#1}>>::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:235:47
  15:     0x7f3f359375b1 - <alloc[a8d04eb2314326e5]::boxed::Box<dyn for<'a, 'b> core[91c473b3610db379]::ops::function::Fn<(&'a std[bebcfa26ff680867]::panic::PanicHookInfo<'b>,), Output = ()> + core[91c473b3610db379]::marker::Send + core[91c473b3610db379]::marker::Sync> as core[91c473b3610db379]::ops::function::Fn<(&std[bebcfa26ff680867]::panic::PanicHookInfo,)>>::call
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/alloc/src/boxed.rs:2220:9
  16:     0x7f3f359375b1 - std[bebcfa26ff680867]::panicking::panic_with_hook
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:833:13
  17:     0x7f3f3590dd7c - std[bebcfa26ff680867]::panicking::panic_handler::{closure#0}
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:691:13
  18:     0x7f3f358fce89 - std[bebcfa26ff680867]::sys::backtrace::__rust_end_short_backtrace::<std[bebcfa26ff680867]::panicking::panic_handler::{closure#0}, !>
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/sys/backtrace.rs:176:18
  19:     0x7f3f3590ff6d - __rustc[9f2c982343335ef0]::rust_begin_unwind
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/panicking.rs:689:5
  20:     0x7f3f313203ec - core[91c473b3610db379]::panicking::panic_fmt
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/panicking.rs:80:14
  21:     0x7f3f328957de - <rustc_errors[63a492c13094a53e]::diagnostic::Diag>::span_suggestion_with_style::<alloc[a8d04eb2314326e5]::string::String, &str>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_errors/src/diagnostic.rs:1030:9
  22:     0x7f3f32c2a24b - <rustc_errors[63a492c13094a53e]::diagnostic::Diag>::span_suggestion::<alloc[a8d04eb2314326e5]::string::String, &str>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_errors/src/diagnostic.rs:1010:14
  23:     0x7f3f32c2a24b - <rustc_hir_analysis[6b4a97736fbac468]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::suggest_removing_args_or_generics::<rustc_span[bfa6808a5e92da49]::ErrorGuaranteed>::{closure#1}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs:1005:17
  24:     0x7f3f32c2a24b - <rustc_hir_analysis[6b4a97736fbac468]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::suggest_removing_args_or_generics::<rustc_span[bfa6808a5e92da49]::ErrorGuaranteed>
  25:     0x7f3f32c2a24b - <rustc_hir_analysis[6b4a97736fbac468]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::suggest::<rustc_span[bfa6808a5e92da49]::ErrorGuaranteed>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs:615:26
  26:     0x7f3f32c2a24b - <rustc_hir_analysis[6b4a97736fbac468]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs as rustc_errors[63a492c13094a53e]::diagnostic::Diagnostic>::into_diag
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs:1151:14
  27:     0x7f3f32ac451c - <rustc_errors[63a492c13094a53e]::DiagCtxtHandle>::create_err::<rustc_hir_analysis[6b4a97736fbac468]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_errors/src/lib.rs:1361:13
  28:     0x7f3f32ac451c - rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::generics::check_generic_arg_count::{closure#2}::{closure#5}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs:557:18
  29:     0x7f3f32ac451c - <core[91c473b3610db379]::option::Option<rustc_span[bfa6808a5e92da49]::ErrorGuaranteed>>::unwrap_or_else::<rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::generics::check_generic_arg_count::{closure#2}::{closure#5}>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/option.rs:1067:21
  30:     0x7f3f32ac451c - rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::generics::check_generic_arg_count::{closure#2}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs:555:43
  31:     0x7f3f32ac451c - rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::generics::check_generic_arg_count
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs:585:9
  32:     0x7f3f32a2eb47 - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_generic_args_of_path::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:527:25
  33:     0x7f3f32a2eb47 - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_generic_args_of_path
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:497:5
  34:     0x7f3f32a191e2 - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_generic_args_of_path_segment
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:456:30
  35:     0x7f3f32a191e2 - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_path_segment
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:1033:25
  36:     0x7f3f32a277e9 - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_resolved_ty_path
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:1991:22
  37:     0x7f3f32a3ab8e - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_ty::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2503:22
  38:     0x7f3f32a3ab8e - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_ty
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2451:5
  39:     0x7f3f32ba473c - <rustc_hir_analysis[6b4a97736fbac468]::collect::ItemCtxt>::lower_ty
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/collect.rs:251:24
  40:     0x7f3f32ba473c - <rustc_hir_analysis[6b4a97736fbac468]::collect::ItemCtxt as rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_fn_sig
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/collect.rs:555:26
  41:     0x7f3f32a0efbc - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2708:43
  42:     0x7f3f32a0efbc - <dyn rustc_hir_analysis[6b4a97736fbac468]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2694:5
  43:     0x7f3f32aa9981 - rustc_hir_analysis[6b4a97736fbac468]::collect::fn_sig::{closure#0}
  44:     0x7f3f32aa9981 - rustc_hir_analysis[6b4a97736fbac468]::collect::fn_sig
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_hir_analysis/src/collect.rs:1019:1
  45:     0x7f3f34719cd5 - rustc_query_impl[23cc7eb0a608f334]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_impl/src/plumbing.rs:302:13
  46:     0x7f3f34719cd5 - rustc_query_impl[23cc7eb0a608f334]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[23cc7eb0a608f334]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_impl/src/plumbing.rs:582:18
  47:     0x7f3f34b7a879 - rustc_query_impl[23cc7eb0a608f334]::query_impl::fn_sig::dynamic_query::{closure#2}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_impl/src/plumbing.rs:659:25
  48:     0x7f3f34b7a879 - <rustc_query_impl[23cc7eb0a608f334]::query_impl::fn_sig::dynamic_query::{closure#2} as core[91c473b3610db379]::ops::function::FnOnce<(rustc_middle[f63169c9694c4243]::ty::context::TyCtxt, rustc_span[bfa6808a5e92da49]::def_id::DefId)>>::call_once
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/ops/function.rs:250:5
  49:     0x7f3f34766f7a - <rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false> as rustc_query_system[db29a4a82fe48170]::query::config::QueryConfig<rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt>>::compute
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_impl/src/lib.rs:119:9
  50:     0x7f3f34766f7a - rustc_query_system[db29a4a82fe48170]::query::plumbing::execute_job_incr::<rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt>::{closure#2}::{closure#2}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_system/src/query/plumbing.rs:569:39
  51:     0x7f3f34766f7a - <rustc_query_system[db29a4a82fe48170]::query::plumbing::execute_job_incr<rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt>::{closure#2}::{closure#2} as core[91c473b3610db379]::ops::function::FnOnce<((rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt, rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false>), rustc_span[bfa6808a5e92da49]::def_id::DefId)>>::call_once
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/ops/function.rs:250:5
  52:     0x7f3f34766f7a - <rustc_query_system[db29a4a82fe48170]::dep_graph::graph::DepGraphData<rustc_middle[f63169c9694c4243]::dep_graph::DepsType>>::with_task::<(rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt, rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false>), rustc_span[bfa6808a5e92da49]::def_id::DefId, rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>::{closure#1}::{closure#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/compiler/rustc_query_system/src/dep_graph/graph.rs:338:64
  53:     0x7f3f34766f7a - rustc_middle[f63169c9694c4243]::ty::context::tls::enter_context::<<rustc_query_system[db29a4a82fe48170]::dep_graph::graph::DepGraphData<rustc_middle[f63169c9694c4243]::dep_graph::DepsType>>::with_task<(rustc_query_impl[23cc7eb0a608f334]::plumbing::QueryCtxt, rustc_query_impl[23cc7eb0a608f334]::DynamicConfig<rustc_query_system[db29a4a82fe48170]::query::caches::DefIdCache<rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>, false, false, false>), rustc_span[bfa6808a5e92da49]::def_id::DefId, rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>::{closure#1}::{closure#0}, rustc_middle[f63169c9694c4243]::query::erase::Erased<[u8; 24usize]>>::{closure#0}
....
....
...
 355:     0x7f1a40f714a4 - std[bebcfa26ff680867]::thread::lifecycle::spawn_unchecked::<rustc_interface[c6c376a60685e945]::util::run_in_thread_with_globals<rustc_interface[c6c376a60685e945]::util::run_in_thread_pool_with_globals<rustc_interface[c6c376a60685e945]::interface::run_compiler<(), rustc_driver_impl[13cf1191ca3affe6]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/thread/lifecycle.rs:90:26
 356:     0x7f1a40f714a4 - <std[bebcfa26ff680867]::thread::lifecycle::spawn_unchecked<rustc_interface[c6c376a60685e945]::util::run_in_thread_with_globals<rustc_interface[c6c376a60685e945]::util::run_in_thread_pool_with_globals<rustc_interface[c6c376a60685e945]::interface::run_compiler<(), rustc_driver_impl[13cf1191ca3affe6]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[91c473b3610db379]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
                               at /rustc-dev/63b1db05801271e400954e41b8600a3cf1482363/library/core/src/ops/function.rs:250:5
 357:     0x7f1a44348b99 - <alloc[a8d04eb2314326e5]::boxed::Box<dyn core[91c473b3610db379]::ops::function::FnOnce<(), Output = ()> + core[91c473b3610db379]::marker::Send> as core[91c473b3610db379]::ops::function::FnOnce<()>>::call_once
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/alloc/src/boxed.rs:2206:9
 358:     0x7f1a44348b99 - <std[bebcfa26ff680867]::sys::thread::unix::Thread>::new::thread_start
                               at /rustc/63b1db05801271e400954e41b8600a3cf1482363/library/std/src/sys/thread/unix.rs:119:17
 359:     0x7f1a3e2969cb - <unknown>
 360:     0x7f1a3e31aa0c - <unknown>
 361:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/matthias/vcs/github/CRED/rustc-ice-2025-12-02T20_39_49-3343166.txt` to your bug report

note: compiler flags: -C incremental=[REDACTED]

query stack during panic:
#0 [fn_sig] computing function signature of `foo`
#1 [crate_variances] computing the variances for items in this crate
... and 4 other queries... use `env RUST_BACKTRACE=1` to see the full query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0412`.
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs
@@ -988,6 +988,9 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                     gen_arg_spans[self.num_expected_type_or_const_args() - 1]
                 };
             let span_hi_redundant_type_or_const_args = gen_arg_spans[gen_arg_spans.len() - 1];
+            if !span_lo_redundant_type_or_const_args.eq_ctxt(span_hi_redundant_type_or_const_args) {
+                return;
+            }
             let span_redundant_type_or_const_args = span_lo_redundant_type_or_const_args
                 .shrink_to_hi()
                 .to(span_hi_redundant_type_or_const_args);

--- a/tests/ui/generics/wrong-number-of-args-in-macro.rs
+++ b/tests/ui/generics/wrong-number-of-args-in-macro.rs
@@ -1,0 +1,15 @@
+// Regression test for #149559
+
+struct Foo<T>; //~ ERROR type parameter `T` is never used
+
+macro_rules! foo_ty {
+    ($a:ty, $b:ty) => {
+        Foo<a, $b>
+        //~^ ERROR cannot find type `a` in this scope
+        //~| ERROR struct takes 1 generic argument but 2 generic arguments were supplied
+    };
+}
+
+fn foo<'a, 'b>() -> foo_ty!(&'b (), &'b ()) {}
+
+fn main() {}

--- a/tests/ui/generics/wrong-number-of-args-in-macro.stderr
+++ b/tests/ui/generics/wrong-number-of-args-in-macro.stderr
@@ -1,0 +1,44 @@
+error[E0425]: cannot find type `a` in this scope
+  --> $DIR/wrong-number-of-args-in-macro.rs:7:13
+   |
+LL |         Foo<a, $b>
+   |             ^ not found in this scope
+...
+LL | fn foo<'a, 'b>() -> foo_ty!(&'b (), &'b ()) {}
+   |                     ----------------------- in this macro invocation
+   |
+   = note: this error originates in the macro `foo_ty` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: you might be missing a type parameter
+   |
+LL | fn foo<'a, 'b, a>() -> foo_ty!(&'b (), &'b ()) {}
+   |              +++
+
+error[E0107]: struct takes 1 generic argument but 2 generic arguments were supplied
+  --> $DIR/wrong-number-of-args-in-macro.rs:7:9
+   |
+LL |         Foo<a, $b>
+   |         ^^^ expected 1 generic argument
+...
+LL | fn foo<'a, 'b>() -> foo_ty!(&'b (), &'b ()) {}
+   |                     ----------------------- in this macro invocation
+   |
+note: struct defined here, with 1 generic parameter: `T`
+  --> $DIR/wrong-number-of-args-in-macro.rs:3:8
+   |
+LL | struct Foo<T>;
+   |        ^^^ -
+   = note: this error originates in the macro `foo_ty` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0392]: type parameter `T` is never used
+  --> $DIR/wrong-number-of-args-in-macro.rs:3:12
+   |
+LL | struct Foo<T>;
+   |            ^ unused type parameter
+   |
+   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
+   = help: if you intended `T` to be a const parameter, use `const T: /* Type */` instead
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0107, E0392, E0425.
+For more information about an error, try `rustc --explain E0107`.

